Fix bugs related to multiple sources

stuarteberg opened the pull request:
After #1929, a single recipe may specify multiple source repos (fixing #1466).  In that case, the metadata `source` section is written as a **list** of source dicts instead of a single dict:

```yaml
source:
  - git_url: https://github.com/foo/foo
    git_tag: 1.0

  - git_url: https://github.com/foo/bar
    git_tag: 1.1
    folder: bar
```

But that breaks an assumption in the convenience function `MetaData.get_value()` -- that every section is a `dict`.  Calls like `meta.get_value('source/git_url')` result in errors.  For example:

```pytb
  File "/miniforge/lib/python3.6/site-packages/conda_build/metadata.py", line 888, in get_value
    value = self.get_section(section).get(key, default)
AttributeError: 'list' object has no attribute 'get'
```

~~This PR "fixes" the problem by merely assuming the caller is interested in the FIRST source in the list.~~

If you don't like this proposal, here are some alternative solutions we could consider:
- ~~Add an assertion to forbid using `MetaData.get_value()` for the `source` section altogether, and require callers to check the type of `meta.get_section('source')` themselves.~~
- Extend `MetaData.get_value()` to support indexes, e.g. `meta.get_value('source/0/git_url')`.

cc: @msarahan (Since you authored #1929)

msarahan commented:
Hmm, sorry about this.  I have found the get_* methods to be annoyingly constrained, and rarely use them.  It shows, I guess. =(

Are you seeing this because you're manipulating MetaData objects yourself in other code?  It's OK if you are, I just want to understand if there's conda-build bugs here, or if I'm only breaking MetaData's API for outside users.

Given your three options, I think a combination of this PR, plus your last suggestion, is nicest.  When a list is present and no index is present in the path, implicitly return only the first element (but show a warning with the index syntax).

stuarteberg commented:
>Are you seeing this because you're manipulating MetaData objects yourself in other code?

No, I was merely trying to write a recipe.  The test case I added in this PR is an example that fails on master.

>I think a combination of this PR, plus your last suggestion, is nicest.

OK, sure.  Now I've updated the PR.

stuarteberg commented:
AFAICT, the Travis failures are unrelated these changes.

For `CONDA_VERSION=4.3.x`, there's just one failure:

```
[gw0] FAILED tests/test_api_build.py::test_disable_pip 
tests/test_api_build.py::test_noarch_none_value 
```

stuarteberg commented:
Regarding the complaint from codacy: Do I really need to avoid using `assert`?!!??

msarahan commented:
RE: codacy, I've been ignoring that, but they have a point.  If those checks are ever skipped, our programs will exit in perhaps strange, uninterpretable ways.  I have that on the way-back burner.

The other stuff looks great.  Thanks for humoring me.  I'm going to go ahead and merge this.  Hopefully I can get the tests all green again soon.

mandeep commented:
Thanks for this. I believe this fixes #2236 as well.
